### PR TITLE
[TwigBridge][TwigBundle][HttpKernel] prefer getSourceContext() over getSource()

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #20435
| License       | MIT
| Doc PR        | n/a

### CODE REVIEW

File: src/Symfony/Bridge/Twig/TwigEngine.php [hunk 1]
```diff
@@ -75,14 +75,14 @@ public function exists($name)
 
         $loader = $this->environment->getLoader();
 
-        if ($loader instanceof \Twig_ExistsLoaderInterface) {
+        if ($loader instanceof \Twig_ExistsLoaderInterface || method_exists($loader, 'exists')) {
```
Comment (xabbuh) [inline src/Symfony/Bridge/Twig/TwigEngine.php]: In Twig 2.0, the `exists()` method will be present in the `Twig_LoaderInterface`.

File: src/Symfony/Component/HttpKernel/Fragment/HIncludeFragmentRenderer.php [hunk 2]
```diff
@@ -140,12 +140,16 @@ private function templateExists($template)
...
             return $loader->exists($template);
         }
 
         try {
-            $loader->getSource($template);
+            if (method_exists($loader, 'getSourceContext')) {
+                $loader->getSourceContext($template);
+            } else {
+                $loader->getSource($template);
+            }
```
Comment (ShinDarth) [inline src/Symfony/Component/HttpKernel/Fragment/HIncludeFragmentRenderer.php]: I would suggest: ``... } else if (method_exists($loader, 'getSource')) { ...``

File: src/Symfony/Bridge/Twig/TwigEngine.php (hunk 1 above)
Comment (xabbuh) [inline src/Symfony/Bridge/Twig/TwigEngine.php]: The call to `exists()` will never be made though if the loader does not implement `Twig_ExistsLoaderInterface`, but is just an instance of `Twig_LoaderInterface`.

Comment (fabpot) [inline src/Symfony/Bridge/Twig/TwigEngine.php]: indeed

Comment (xabbuh) [inline src/Symfony/Bridge/Twig/TwigEngine.php]: And this seems to happen. Otherwise, I don't see how #20435 would ever reach the `getSource()` call.

**Comment by xabbuh** [comment]
fabbot failures are unrelated.

**Comment by fabpot** [comment]
Thank you @xabbuh.

**Comment by ShinDarth** [comment]
good job @xabbuh

**Pull request merged**